langchain/src/chat_models/base.ts

```diff
@@ -292,6 +316,83 @@ export abstract class BaseChatModel<
     return output;
   }
 
+  /**
+   * Generates chat based on the input messages.
+   * @param messages An array of arrays of BaseMessage instances.
+   * @param options The call options or an array of stop sequences.
+   * @param callbacks The callbacks for the language model.
+   * @returns A Promise that resolves to an LLMResult.
+   */
+  async generate(
+    messages: BaseMessageLike[][],
+    options?: string[] | CallOptions,
+    callbacks?: Callbacks
+  ): Promise<LLMResult> {
+    // parse call options
+    let parsedOptions: CallOptions | undefined;
+    if (Array.isArray(options)) {
+      parsedOptions = { stop: options } as CallOptions;
+    } else {
+      parsedOptions = options;
+    }
+
+    const baseMessages = messages.map((messageList) =>
+      messageList.map(coerceMessageLikeToMessage)
+    );
+
+    const [runnableConfig, callOptions] =
+      this._separateRunnableConfigFromCallOptions(parsedOptions);
+    runnableConfig.callbacks = runnableConfig.callbacks ?? callbacks;
+
+    if (!this.cache) {
+      return this._generateUncached(baseMessages, callOptions, callbacks);
+    }
+
+    const { cache } = this;
+    const params = this.serialize();
+    params.stop = callOptions.stop ?? params.stop;
+
+    const llmStringKey = `${Object.entries(params).sort()}`;
```

Comment:
I think this needs to include the entire call options too

Reply:
Refactored to add call options, however for some context I did it this way because that's how it's being done in `llms/base` ([see link](https://github.com/langchain-ai/langchainjs/blob/6e1a112934c2bbeb0bcc610c1a8a5cd2c685c1c7/langchain/src/llms/base.ts#L326C8-L326C8))
```typescript
    const { cache } = this;
    const params = this.serialize();
    params.stop = callOptions.stop ?? params.stop;

    const llmStringKey = `${Object.entries(params).sort()}`;
```

If you think that's also wrong, I can add that refactor too (I think that means it would break users existing cache)

Reply:
Nuno is right (and this is specifically important for chat models since "functions" are a thing), and could become important for LLMs in the near future as well.

I moved the cache key creation check into `BaseLanguageModel` - I think impact here will be pretty limited (caching with no TTL isn't generally a great idea anyway) and we can announce the change.